metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -139,15 +147,47 @@ class CheckpointingService[F[_]: Sync: Applicative, N] private (
       ledgerStorage.put(ledger) >>
         viewStateStorage.setLastExecutedBlockHash(header.hash)
     } >>
-      ledgerTree.set(Map(header.hash -> (ledger, header))) >>
+      ledgerTree.set(Map(header.hash -> (ledger, header, None))) >>
       lastExecutedHeader.set(header)
   }
 
-  private def constructCheckpoint(
-      ledger: Ledger,
+  private def constructCheckpointCert(
+      commitHeader: Block.Header,
       commitQC: QuorumCertificate[CheckpointingAgreement]
-  ): F[Option[CheckpointCertificate]] =
-    ??? //TODO: PM-3110
+  ): F[Option[CheckpointCertificate]] = {
+
+    def traverseCommitPath(
+        ledgers: LedgerTree,
+        from: Block.Hash,
+        path: NonEmptyList[Block.Header]
+    ): Option[CheckpointCertificate] =
+      ledgers.get(from) match {
+        case Some((_, _, Some(cp))) =>
+          CheckpointCertificate(path, cp.checkpoint, cp.proof, commitQC).some
+        case Some((_, h, None)) =>
+          traverseCommitPath(ledgers, h.parentHash, path :+ h)
+        case None => None
+      }
+
+    ledgerTree.get.map(
+      traverseCommitPath(_, commitQC.blockHash, NonEmptyList.one(commitHeader))
+    )
+  }
+
+  //TODO: could be optimised for no checkpoint case
+  //TODO: avoid .get
+  private def contructCheckpointProof(
+      block: Block,
+      ledger: Ledger
+  ): Option[CheckpointWithProof] = {
+    val tree = MerkleTree.build(
+      block.body.transactions.map(tx => MerkleTree.Hash(tx.hash))
+    )
+    val checkpoint     = ledger.maybeLastCheckpoint.get
```

Comment:
You can avoid the `.get` by just finding the last checkpoint in the block body. Arguably there's nothing to prove if there was no checkpoint added in the block, so constructing a Merkle tree and then trying to find the checkpoint in the ledger that could be from a previous state is more work than finding the the checkpoint and constructing a proof from it. Shouldn't even need the `ledger` for this, unless as a sanity check that the last checkpoint in the block matches the ledger.

Also, there may not be a `maybeLastCheckpoint` in the ledger yet, in case we just added the first proposer block, or we're just committing an empty block after genesis. We can't start with a checkpoint because we have no idea what's on the PoW side.

Reply:
What if we had a special value for genesis so that we can avoid the `Option`, like:

```
val GenesisCheckpointCandidate = CheckpointCandidate(BitVector.empty)
```
?

I realise that building the proof is not optimal, as more often it will be the case that we don't have a checkpoint in the block. It was just nicer to code this way, but I will update this.

Reply:
If we use `BitVector.empty` to indicate that this is a genesis, then I guess the problem is that 
1. We assume that the checkpointing interpreter doesn't itself want to use `BitVector.empty`, whereas so far we said we don't try to guess what's in a checkpoint.
2. We exchange a compiler supported condition of having no checkpoint so far with another kind of `.isEmpty`, but one we can forget. For example here you would get rid of a `.get`, but we would be sending the `Ledger` in all validation and block creation requests, and want the interpreter implementer be aware while checking for example that we are extending the previous checkpoint in the ledger that it may not be a checkpoint after all. At least we probably can't produce an invalid certificate because that would require the body to have one like this as well, which would hopefully fail to validate.

I thought turning `Ledger` into an ADT would be an overkill, but that would be another option, that we are either before we got our first checkpoint, or after.

Reply:
No longer using `.get` so that part's resolved.

But I'm still thinking we could use a `GenesisCheckpointCandidate`. Its value could be agreed upon via genesis configuration. This is similar to how we define a genesis block. We could also have `parentHash: Option[Hash]`, but that would be cumbersome in the long run, so we specify some artificial value. The PoW interpreter could propose such a value and we'd include it in the definition of the genesis block in metronome.

Reply:
I'm not against putting the identity of the PoW genesis block into the starting ledger as the first checkpoint, but I'd think about extending the RPC interface where we can query the interpreter for this information upon startup, rather than try to manage it manually in configuration. 

@dzajkowski already asked if the checkpoints can have the block height as well as the identity in them, and sure enough, the interpreters can encode any content into the binary values inside the checkpoint. A height of 0 could indicate that this is the genesis candidate, and the interpreter can use this information to validate the block in a different way.

Because of that, I don't know why we would need a separate data type like a `GenesisCheckpointCandidate`. 

I'm also not sure this brings much to the table in terms of extra knowledge that the interpreter wouldn't have if we don't add it: in order to trust this value, it must come from the interpreter, and inside the interpreter it can always be used where we had `None` so far. The only spot where it provides something new that you can immediately tell that you the other federation members are on a different chain, if the hashes don't match, so you cannot extend it with a new checkpoint. But this can only happen at the beginning, after that they move on, and you just stand there not knowing the checkpoints they are talking about at higher levels.

Reply:
`GenesisCheckpointCandidate` - I don't know why I wrote it like that. I meant just the concept of having a genesis value, it certainly doesn't need to be expressed by another type. And I like the idea of querying the interpreter for it, rather than relying on config :+1: